File: langchain-integration/src/security.py
"""Allowlisted evidence validation and privacy scans."""

from __future__ import annotations

import re
from typing import Any


FORBIDDEN_RESULT_FIELDS = {
    "api_key",
    "authorization",
    "headers",
    "prompt",
    "messages",
    "output",
    "content",
    "reasoning_content",
    "request_id",
    "provider_request_id",
    "tool_call_id",
    "arguments",
    "retrieved_context",
    "raw",
    "balance",
    "account",
}
SECRET_PATTERNS = (
    re.compile(r"\bsk-[A-Za-z0-9_-]{16,}\b"),
    re.compile(r"\bBearer\s+[A-Za-z0-9._~+/-]{12,}=*\b", re.IGNORECASE),
    re.compile(
        r"(?:api[_-]?key|authorization)\s*[:=]\s*[\"'][^\"']{8,}[\"']",
        re.IGNORECASE,
    ),
)
NON_ASCII_PATTERN = re.compile(r"[^\x00-\x7f]")
MOJIBAKE_TOKENS = (
    "\u00c3",
    "\u00c2",
    "\u00e2\u20ac",
    "\u00f0\u0178",
    "\u00ef\u00bb\u00bf",
    "\ufffd",
)


def assert_allowlisted_result(value: Any, path: str = "$") -> None:
    if isinstance(value, list):
        for index, item in enumerate(value):
            assert_allowlisted_result(item, f"{path}[{index}]")
        return
    if not isinstance(value, dict):
        return
    for key, child in value.items():
        if key.lower() in FORBIDDEN_RESULT_FIELDS:
            raise ValueError(f"Forbidden result field at {path}.{key}")
        assert_allowlisted_result(child, f"{path}.{key}")


def secret_findings(text: str) -> int:
    findings = 0
    for pattern in SECRET_PATTERNS:
        for match in pattern.finditer(text):
            if any(
                marker in match.group(0)
                for marker in (
                    "offline-only-not-a-credential",
                    "set-outside-source-control",
                )
            ):
                continue
            findings += 1
    return findings


def text_findings(text: str) -> dict[str, int]:
    return {
        "non_ascii_characters": len(NON_ASCII_PATTERN.findall(text)),
        "mojibake_matches": sum(text.count(token) for token in MOJIBAKE_TOKENS),
    }


def safe_error_class(error: BaseException) -> str:
    name = error.__class__.__name__
    return name if re.fullmatch(r"[A-Za-z][A-Za-z0-9_]{0,79}", name) else "Error"


def safe_error_code(error: BaseException) -> str | None:
    code = getattr(error, "code", None)
    if isinstance(code, str) and re.fullmatch(r"[A-Za-z0-9_.-]{1,80}", code):
        return code
    return None


def safe_status(error: BaseException) -> int | None:
    for name in ("status_code", "status"):
        value = getattr(error, name, None)
        if isinstance(value, int) and 100 <= value <= 599:
            return value
    return None

File: nodejs-typescript/src/security.ts
const FORBIDDEN_RESULT_FIELDS = new Set([
  "api_key",
  "authorization",
  "headers",
  "cookies",
  "prompt",
  "messages",
  "output",
  "content",
  "reasoning_content",
  "request_id",
  "_request_id",
  "response_id",
  "provider_request_id",
  "tool_call_id",
  "arguments",
  "tool_result",
  "retrieved_context",
  "raw",
  "stack",
  "message",
  "balance",
  "account",
  "user",
]);

const SECRET_PATTERNS = [
  /\bsk-[A-Za-z0-9_-]{16,}\b/g,
  /\bBearer\s+[A-Za-z0-9._~+/-]{12,}=*\b/gi,
  /(?:api[_-]?key|authorization)\s*[:=]\s*["'][^"']{8,}["']/gi,
];

const MOJIBAKE_MARKERS = ["\u00c3", "\u00c2", "\u00e2\u20ac", "\ufffd"];

export interface TextFindings {
  secret_findings: number;
  non_ascii_characters: number;
  mojibake_matches: number;
}

export function assertAllowlistedResult(value: unknown, path = "$"): void {
  if (Array.isArray(value)) {
    value.forEach((item, index) => assertAllowlistedResult(item, `${path}[${index}]`));
    return;
  }
  if (!value || typeof value !== "object") {
    return;
  }
  for (const [key, child] of Object.entries(value)) {
    if (FORBIDDEN_RESULT_FIELDS.has(key.toLowerCase())) {
      throw new Error(`Forbidden evidence field at ${path}.${key}.`);
    }
    assertAllowlistedResult(child, `${path}.${key}`);
  }
}

export function inspectText(text: string): TextFindings {
  let secretFindings = 0;
  for (const pattern of SECRET_PATTERNS) {
    pattern.lastIndex = 0;
    secretFindings += [...text.matchAll(pattern)].length;
  }
  return {
    secret_findings: secretFindings,
    non_ascii_characters: [...text].filter((character) => character.codePointAt(0)! > 127)
      .length,
    mojibake_matches: MOJIBAKE_MARKERS.reduce(
      (count, marker) => count + text.split(marker).length - 1,
      0,
    ),
  };
}

export function assertSafeEvidence(value: unknown): void {
  assertAllowlistedResult(value);
  const encoded = JSON.stringify(value);
  const findings = inspectText(encoded);
  if (
    findings.secret_findings !== 0 ||
    findings.non_ascii_characters !== 0 ||
    findings.mojibake_matches !== 0
  ) {
    throw new Error("Evidence text failed the privacy scan.");
  }
}

export function safeErrorClass(error: unknown): string {
  const candidate =
    typeof error === "object" &&
    error !== null &&
    "constructor" in error &&
    typeof error.constructor === "function"
      ? error.constructor.name
      : "Error";
  return /^[A-Za-z][A-Za-z0-9_]{0,79}$/.test(candidate) ? candidate : "Error";
}

export function safeStatus(error: unknown): number | null {
  if (
    typeof error === "object" &&
    error !== null &&
    "status" in error &&
    Number.isInteger(error.status) &&
    Number(error.status) >= 100 &&
    Number(error.status) <= 599
  ) {
    return Number(error.status);
  }
  return null;
}

export function safeErrorCode(error: unknown): string | null {
  if (
    typeof error === "object" &&
    error !== null &&
    "code" in error &&
    typeof error.code === "string" &&
    /^[A-Za-z0-9_.-]{1,80}$/.test(error.code)
  ) {
    return error.code;
  }
  return null;
}

export function safeToken(value: unknown, maximum = 100): string | null {
  return typeof value === "string" &&
    value.length <= maximum &&
    /^[A-Za-z0-9._-]+$/.test(value)
    ? value
    : null;
}
